### DISCUSSION TITLE
Multiple monitors with Aerospace + SketchyBar

### DISCUSSION THREAD

**Discussion by alcidesmig** [FelixKratz/SketchyBar#624] [Q&A]
Hello everyone!

I used to use yabai + sketchybar, but I'm migrating to aerospace + sketchybar.
In a multi-monitor setup, I'm having a hard time getting the desired behavior. I'd like each monitor to only display the workspaces associated with it. I've tried a few approaches, like multiple bars, items associated to specific displays, but I haven't managed to get anything to look good.

Currently what I have is that all the bars appear on all monitors and when I change the workspace in one display, the change (i.e., the highlighted workspace in the bar) happens on all monitors (leading to wrong highlighting on displays other than the one where the change happened).

The important sections of my config files are:

```
# ~/.config/sketchybar/sketchybarrc
sketchybar --add event aerospace_workspace_change

for sid in $(aerospace list-workspaces --all); do
    sketchybar --add item space.$sid left \
        --subscribe space.$sid aerospace_workspace_change \
        --set space.$sid \
        background.color=0x44ffffff \
        background.corner_radius=20 \
        background.height=20 \
	background.border_color=0x00FFFF \
	padding_left=7 \
	padding_right=7 \
        background.drawing=off \
        label="$sid" \
        click_script="aerospace workspace $sid" \
        script="$CONFIG_DIR/plugins/aerospace.sh $sid"
done
```

```
# ~/.aerospace.toml
exec-on-workspace-change = ['/bin/bash', '-c',
    'sketchybar --trigger aerospace_workspace_change FOCUSED_WORKSPACE=$AEROSPACE_FOCUSED_WORKSPACE'
]

[workspace-to-monitor-force-assignment]
1 = 'main'
2 = 'main'
3 = 'main'
4 = '27GL'
5 = '27GL'
6 = '27GL'
7 = '2718'
8 = '2718'
9 = '2718'
```

Thanks in advance.

**Comment by kaynetik** [comment]
Up! 🚀

Actually facing the same conundrum at the moment as well.

**Comment by jdurand** [comment]
I'm encountering the same issue and currently relying on a workaround in my SketchyBar config, which is far from ideal and prone to breaking. Here's the function I use in my setup using [SbarLua](https://github.com/FelixKratz/SbarLua):

```lua
-- Map monitor IDs between AeroSpace and SketchyBar
local function sketchyBarMonitorId(monitor_id, total_monitors)
  return (total_monitors > 1 and monitor_id == 1) and 2 or (monitor_id == 2 and 1 or monitor_id)
end
```

A good workaournd would be to have a SketchyBar command that directly outputs a monitor ID-to-name mapping. Similar to what AeroSpace provides:
```
$ aerospace list-monitors --json
[
  {
    "monitor-name" : "VP2771",
    "monitor-id" : 1
  },
  {
    "monitor-id" : 2,
    "monitor-name" : "Built-in Retina Display"
  }
]
```

**Comment by wesharper** [comment]
It's not super ideal, but I was hacking at this and came up with a few workarounds here: https://github.com/FelixKratz/SketchyBar/issues/607

**Comment by Masstronaut** [comment]
I have been living with a similar setup for a while (all workspaces are displayed on all sketchybars). A couple days ago I decided to fix it so that the sketchybar on each display only renders the aerospace workspaces associated with that display.

The commit where I implemented it is [here](https://github.com/Masstronaut/dotfiles/commit/8cb79db1101968d70d848700e438d0026438b72e). 

My solution works well and handles:
- Front app is only displayed on the monitor that houses it
- correctly maps between aerospace monitors and NSScreen screen IDs (which are used by sketchybar)
- When a workspace is moved from one monitor to another, the sketchybar item for that workspace is also moved

The crux of it is that aerospace lets you format the output of queries and so they can be formatted to return the NSScreen screen ID for anything (workspace, window, etc.) so I use that to determine which sketchybar monitor to render items on. Here's the code for that:

```sh
# Get the system monitor ID of the specified workspace. 
# Expects the workspace ID as the first argument
# Returns the monitor ID of the workspace
# returns nothing (empty string) if the workspace has no windows.
get_workspace_monitor_id() {
  local workspace_id="$1"
  # aerospace command that returns all windows in the workspace, formatted as only the system monitor ID
  # each workspace is 1 character long, so truncate the output to only the first character
  local aerospace_result=$(aerospace list-windows --workspace $workspace_id --format "%{monitor-appkit-nsscreen-screens-id}")
  local monitor_id=${aerospace_result:0:1}
  echo $monitor_id
}
```

and consuming the `get_workspace_monitor_id` function
```sh
# monitor_id will contain the monitor ID of the workspace
# or an empty string if the workspace has no windows
local monitor_id=$(get_workspace_monitor_id $sid)
```

The one case I haven't handled which comes up fairly often for me is when the number of attached displays changes. I work on a laptop and sometimes connect an external display, sometimes not. When I transition from a display being connected to not connected, the sketchybar items on the disconnected display don't get moved to the built-in display.

If anyone knows how to listen for an event that the number of connected displays changed, I'd be really grateful for info about how to do that.

**Reply by timharris777** [reply]
I have not been able to listen for an event for display added or removed. I'm currently handling this with the following in my aerospace config:
```
...
after-startup-command = [
	'exec-and-forget sketchybar --reload',
	'exec-and-forget borders'
]
...
[mode.service.binding]
  esc = ['exec-and-forget killall AeroSpace && open /Applications/AeroSpace.app || open /Applications/AeroSpace.app']
...
```
Anytime anything is not correct in sketchybar or aerospace I run `opt+shft+;` and press `esc` which kills and restarts aerospace and in the process runs 'sketchybar --reload`.

**Comment by anatoli-tsinovoy** [comment]
Hi! I've found that for my setup there appear to be four numberings that need to be taken into consideration.

1. The aerospace monitors (they have their own numbering according to aerospace arrangement rules)
2. The sketchybar monitors (what we interact with using `--set space space.1 display=2` etc.)
3. The numbering according to AppKit's NSScreen
4. The numbering according to CoreGraphics (DirectDisplayID)

I tested my setup with a single laptop monitor, laptop monitor+two external displays, two external displays with closed-lid laptop.

Here are my outputs for the four numbering systems for two of the setups, as far as I can tell, all of the added complexity here is warranted.

Two external monitors, laptop lid closed:
```
| NSScreen | Direct | Aerospace | Sketchybar |
|----------|--------|-----------|------------|
| 1        | 3      | 1         | 1          |
| 2        | 2      | 2         | 2          |
```

Two external monitors, laptop lid open:
```
| NSScreen | Direct | Aerospace | Sketchybar |
|----------|--------|-----------|------------|
| 1        | 1      | 2         | 1          |
| 2        | 2      | 3         | 3          |
| 3        | 3      | 1         | 2          |
```

Note how the NSScreen numbers and the sketchybar numbers don't match up in this case, so the common approach of using NSScreen as the sketchybar display number does not work in this case.


```bash
#!/bin/bash

while IFS=" " read -r nsscreen_id direct_id; do
  NSSCREEN_TO_DIRECT[$nsscreen_id]="$direct_id"
done < <(swift -e 'import AppKit; import CoreGraphics; for (i, scr) in NSScreen.screens.enumerated() {
  let n = scr.deviceDescription[NSDeviceDescriptionKey("NSScreenNumber")] as! NSNumber
  if let u = CGDisplayCreateUUIDFromDisplayID(CGDirectDisplayID(truncating: n))?.takeRetainedValue() {
    // let s = CFUUIDCreateString(nil, u) as truncating
    // print("\(i+1) \(n.intValue) \(s)")
    print("\(i+1) \(n.intValue)")
  }
}')

while IFS=" " read -r direct_id sb_monitor; do
  DIRECT_TO_SB[$direct_id]="$sb_monitor"
done < <(sketchybar --query displays | jq -r '.[] | "\(.["DirectDisplayID"]) \(.["arrangement-id"])"')

while IFS=" " read -r as_monitor nsscreen_id; do
  AS_TO_SB[$as_monitor]=${DIRECT_TO_SB["${NSSCREEN_TO_DIRECT["-$nsscreen_id"]}"]}
done < <(aerospace list-monitors --format '%{monitor-id} %{monitor-appkit-nsscreen-screens-id}')

if ! $(sketchybar --query AS_TO_SB &>/dev/null); then
  sketchybar --add item AS_TO_SB left --set AS_TO_SB display=0 label="${AS_TO_SB[*]}"
fi

```

**Reply by fbuetler** [reply]
As @anatoli-tsinovoy correctly pointed out, both aerospace and sketchybar have an internal numbering of the monitors:

>     1. The aerospace monitors (they have their own numbering according to aerospace arrangement rules)
>     2. The sketchybar monitors (what we interact with using `--set space space.1 display=2` etc.)
>     3. The numbering according to AppKit's NSScreen
>     4. The numbering according to CoreGraphics (DirectDisplayID)

Further, 
* aerospace uses NSScreen to display the workspace
* sketchybar uses DirectDisplay to display the item

So, we need a mapping from NSScreen to DirectDisplay (the internal numbering of both aerospace and sketchybar is not important in that regard). 

The Swift code snippet above essentially does that.
With that we can make the mapping `workspace (monitor-id) -> NSScreen -> DirectScreen -> item (arrangement-id)`.

However, I noticed that the Swift code is pretty slow, so I am experiencing a delay from the point I switch an aerospace workspace to until sketchybar highlights the respective item. I think some caching is required in this regard, as the mapping should (?) not change, while the number of monitors stays the same.

**Reply by anatoli-tsinovoy** [reply]
The swift -e call essentially compiles and runs the snippet on every call, so it makes sense for it to be a slow call.

On my setup I cache it as the label of a display=0 sketchybar item, so I don't call the code often.

I also piped it to restart sketchybar whenever a monitor change is detected (though this is a stub for now until I setup a monitor-change-monitor to fire the event)

( https://github.com/anatoli-tsinovoy/dotfiles )

Note I have some nonsense in the comments there about 1-based indexing, it's not true and just me having a hard time understanding arrays in bash3.2

**Reply by Tres0rsHoang** [reply]
I tried to use your idea to map AppKit’s NSScreenID to DirectDisplayID, and I realized that this command in Aerospace is pretty useful:
(aerospace list-monitors --format "%{monitor-appkit-nsscreen-screens-id}") → this already maps the AppKit NSScreenID to the corresponding DirectDisplayID. The only thing we need is to loop and show each workspace on the correct monitor.

```
#!/bin/bash

count=1;
for ws in $(aerospace list-monitors --format "%{monitor-appkit-nsscreen-screens-id}"); do
  for sid in $(aerospace list-workspaces --monitor $count); do
    sketchybar --set space.$sid display=$ws
  done
  count=$((count + 1))
done
```

**Reply by fbuetler** [reply]
Your command returns the NSScreen Ids. These are not the same as the DirectScreen ID. It might match, but is not guaranteed to. See this example (with 2 external monitors and the laptop lid closed), where NSScreenID != DirectDisplayID:

```bash
➜ aerospace list-monitors --format "%{monitor-id} %{monitor-appkit-nsscreen-screens-id}"
1 2
2 1

➜ swift -e 'import AppKit; import CoreGraphics; for (i, scr) in NSScreen.screens.enumerated() {
  let n = scr.deviceDescription[NSDeviceDescriptionKey("NSScreenNumber")] as! NSNumber
  print("\(i+1) \(n.intValue)")
}'
1 3
2 2

➜ sketchybar --query displays | jq -r '.[] | "\(.["DirectDisplayID"]) \(.["arrangement-id"])"'
3 1
2 2
```

**Reply by obsidian33** [reply]
@fbuetler did you ever find better fix/workaround? You mentioned implementing some caching to help with the swift command. Do you have a good solution for this?

**Comment by haxybaxy** [comment]
https://github.com/nikitabobko/AeroSpace/issues/1656 I made an issue for adding a main monitor flag to the various commands, so you that you don't have to depend on the numbers of the monitors which change based on the position. I think this will solve a lot of problems for dual monitor set ups, and hopefully will help a bit if you have more than two
